io/private/EptSupport.hpp

```diff
@@ -90,6 +158,127 @@ inline std::array<double, 3> toArray3(const Json::Value& json)
     return result;
 }
 
+class PDAL_DLL Key
+{
+    // An EPT key representation (see https://git.io/fAiBh).  A depth/X/Y/Z key
+    // representing a data node, as well as the bounds of the contained data.
+public:
+    Key(BOX3D b) : b(b) { }
+    Key(std::string s)
+    {
+        const std::vector<std::string> tokens(Utils::split(s, '-'));
+        if (tokens.size() != 4)
+            throw std::runtime_error("Invalid EPT KEY: " + s);
+        d = std::stoull(tokens[0]);
+        x = std::stoull(tokens[1]);
+        y = std::stoull(tokens[2]);
+        z = std::stoull(tokens[3]);
+    }
+
+    BOX3D b;
+    uint64_t d = 0;
+    uint64_t x = 0;
+    uint64_t y = 0;
+    uint64_t z = 0;
+
+    std::string toString() const
+    {
+        return std::to_string(d) + '-' + std::to_string(x) + '-' +
+            std::to_string(y) + '-' + std::to_string(z);
+    }
+
+    double& operator[](uint64_t i)
+    {
+        switch (i)
+        {
+            case 0: return b.minx;
+            case 1: return b.miny;
+            case 2: return b.minz;
+            case 3: return b.maxx;
+            case 4: return b.maxy;
+            case 5: return b.maxz;
+            default: throw pdal_error("Invalid Key[] index");
+        }
+    }
+
+    uint64_t& idAt(uint64_t i)
+    {
+        switch (i)
+        {
+            case 0: return x;
+            case 1: return y;
+            case 2: return z;
+            default: throw pdal_error("Invalid Key::idAt index");
+        }
+    }
+
+    Key bisect(uint64_t direction) const
+    {
+        Key key(*this);
+        ++key.d;
+
+        auto step([&key, direction](uint8_t i)
+        {
+            key.idAt(i) *= 2;
+
+            const double mid(key[i] + (key[i + 3] - key[i]) / 2.0);
+            const bool positive(direction & (((uint64_t)1) << i));
+            if (positive)
+            {
+                key[i] = mid;
+                ++key.idAt(i);
+            }
+            else
+            {
+                key[i + 3] = mid;
+            }
+        });
+
+        for (uint8_t i(0); i < 3; ++i)
+            step(i);
+
+        return key;
+    }
+};
+
```

Comment:
Do did the Key stuff just move in this file and the diff engine couldn't splice it?  Did something else change?

Reply:
The Key stuff probably just shifted a bit since some utils formerly in the EptReader unnamed namespace got moved here for reuse.  The only thing that changed in the `Key` class was the addition of a string constructor, used by the EptAddonWriter.